src/stream/src/executor/backfill/snapshot_backfill/vnode_stream.rs

```diff
@@ -0,0 +1,200 @@
+                None => Ok(None),
+                Some((None, stream)) => {
+                    assert!(self.finished_vnode.insert(stream.vnode));
+                    continue;
+                }
+                Some((Some(Ok(item)), stream)) => {
+                    self.streams.push(stream.into_future());
+                    Ok(Some(item))
+                }
+                Some((Some(Err(e)), _stream)) => Err(e),
+            };
+            break Poll::Ready(ready_item);
+        }
+    }
+
+    #[expect(dead_code)]
+    pub(super) fn consume_builder(&mut self) -> Option<StreamChunk> {
+        self.data_chunk_builder.consume_all().map(|chunk| {
+            let ops = replace(
+                &mut self.ops,
+                Vec::with_capacity(self.data_chunk_builder.batch_size()),
+            );
+            StreamChunk::from_parts(ops, chunk)
+        })
+    }
+
+    #[expect(dead_code)]
+    pub(super) async fn for_vnode_pk_progress(
+        &mut self,
+        pk_indices: &[usize],
+        mut on_vnode_progress: impl FnMut(VirtualNode, Option<OwnedRow>),
+    ) -> StreamExecutorResult<()> {
+        assert!(self.data_chunk_builder.is_empty());
+        for vnode in &self.finished_vnode {
+            on_vnode_progress(*vnode, None);
+        }
+        for vnode_stream in &mut self.streams {
+            let vnode_stream = vnode_stream.get_mut().expect("should exist");
+            match vnode_stream.stream.as_mut().peek().await {
+                Some(Ok(((_, row), second))) => {
+                    let pk = row.project(pk_indices).to_owned_row();
+                    if cfg!(debug_assertions)
+                        && let Some((_, second_row)) = second
+                    {
+                        assert_eq!(pk, second_row.project(pk_indices).to_owned_row());
+                    }
+                    on_vnode_progress(vnode_stream.vnode, Some(pk));
+                }
+                Some(Err(_)) => {
+                    return Err(vnode_stream
+                        .stream
+                        .try_next()
+                        .await
+                        .expect_err("checked Err"));
+                }
+                None => {
+                    on_vnode_progress(vnode_stream.vnode, None);
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<St: BackfillRowStream> Stream for VnodeStream<St> {
+    type Item = StreamExecutorResult<StreamChunk>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+        let capacity = this.data_chunk_builder.batch_size();
+        loop {
+            match ready!(this.poll_next_row(cx)) {
+                Ok(Some(((op, row), second))) => {
+                    let may_chunk = if let Some((second_op, second_row)) = second {
+                        if this.data_chunk_builder.can_append(2) {
```

Comment:
```suggestion
                        if this.data_chunk_builder.can_append_update
```

Two rows will always correspond to update. This will make it more readable.